emu_sv/sv_backend_impl.py

```diff
@@ -0,0 +1,264 @@
+    ):
+        super().__init__(evaluation_times=evaluation_times)
+        self.data = data
+        self.timestep_count = timestep_count
+
+    @property
+    def _base_tag(self) -> str:
+        return "statistics"
+
+    def apply(
+        self,
+        *,
+        config: EmulationConfig,
+        state: State,
+        **kwargs: typing.Any,
+    ) -> dict:
+        """Calculates the observable to store in the Results."""
+        assert isinstance(state, StateVector | DensityMatrix)
+        assert isinstance(config, SVConfig)
+        duration = self.data[-1]
+        if isinstance(state, StateVector) and state.vector.is_cuda:
+            max_mem_per_device = (
+                torch.cuda.max_memory_allocated(device) * 1e-6
+                for device in range(torch.cuda.device_count())
+            )
+            max_mem = max(max_mem_per_device)
+        elif isinstance(state, DensityMatrix) and state.matrix.is_cuda:
+            max_mem_per_device = (
+                torch.cuda.max_memory_allocated(device) * 1e-6
+                for device in range(torch.cuda.device_count())
+            )
+            max_mem = max(max_mem_per_device)
+        else:
+            max_mem = getrusage(RUSAGE_SELF).ru_maxrss * 1e-3
+
+        config.logger.info(
+            f"step = {len(self.data)}/{self.timestep_count}, "
+            + f"RSS = {max_mem:.3f} MB, "
+            + f"Δt = {duration:.3f} s"
+        )
+
+        return {
+            "RSS": max_mem,
+            "duration": duration,
+        }
+
+
+_TIME_CONVERSION_COEFF = 0.001  # Omega and delta are given in rad/μs, dt in ns
+
+
+class SVBackendImpl:
+    results: Results
+
+    def __init__(self, config: SVConfig, pulser_data: PulserData):
+        """
+        Initializes the SVBackendImpl.
+
+        Args:
+            config: The configuration for the emulator.
+            pulser_data: The data for the sequence to be emulated.
+        """
+        self._config = config
+        self._pulser_data = pulser_data
+        self.target_times = pulser_data.target_times
+        self.omega = pulser_data.omega
+        self.delta = pulser_data.delta
+        self.phi = pulser_data.phi
+        self.nsteps = pulser_data.omega.shape[0]
+        self.nqubits = pulser_data.omega.shape[1]
+
+        self.time = time.time()
+
+        self.results = Results(atom_order=(), total_duration=self.target_times[-1])
+        self.statistics = Statistics(
+            evaluation_times=[t / self.target_times[-1] for t in self.target_times],
+            data=[],
+            timestep_count=self.nsteps,
+        )
+
+    def initial_state(self) -> State:
+        if self._config.initial_state is not None:
+            state = self._config.initial_state
+            return StateVector(state.vector.clone(), gpu=state.vector.is_cuda)
+        else:
+            return StateVector.make(self.nqubits, gpu=self._config.gpu)
+
+    def _choose_stepper(self, state: State) -> typing.Callable:
```

Comment:
The body of this function is the symptom that there is something wrong with your design: `SVBackendImpl` shouldn't mention `EvolveDensityMatrix` at all.

See my comment below where I suggest creating a `Base`

Reply:
Yes, `_choose_stepper` should always return `EvolveStateVector` in the non-noisy class. In the noisy class, you can reimplement this method and always return `EvolveDensityMatrix.evolve`, and then they become nicely separated.